lightning/src/util/ser_macros.rs

```diff
@@ -7,82 +7,142 @@
 // You may not use this file except in accordance with one or both of these
 // licenses.
 
+macro_rules! encode_a_tlv {
+	($stream: expr, $type: expr, $field: expr, explicit) => {
+		BigSize($type).write($stream)?;
+		BigSize($field.serialized_length() as u64).write($stream)?;
+		$field.write($stream)?;
+	};
+	($stream: expr, $type: expr, $field: expr, vec_type) => {
+		encode_a_tlv!($stream, $type, ::util::ser::VecWriteWrapper(&$field), explicit);
+	};
+	($stream: expr, $optional_type: expr, $optional_field: expr, option) => {
+		if let Some(ref field) = $optional_field {
+			BigSize($optional_type).write($stream)?;
+			BigSize(field.serialized_length() as u64).write($stream)?;
+			field.write($stream)?;
+		}
+	};
+}
+
 macro_rules! encode_tlv {
-	($stream: expr, {$(($type: expr, $field: expr)),*}, {$(($optional_type: expr, $optional_field: expr)),*}) => { {
+	($stream: expr, {$(($type: expr, $field: expr, $fieldty: ident)),*}) => { {
 		#[allow(unused_imports)]
 		use util::ser::BigSize;
-		// Fields must be serialized in order, so we have to potentially switch between optional
-		// fields and normal fields while serializing. Thus, we end up having to loop over the type
-		// counts.
-		// Sadly, while LLVM does appear smart enough to make `max_field` a constant, it appears to
-		// refuse to unroll the loop. If we have enough entries that this is slow we can revisit
-		// this design in the future.
-		#[allow(unused_mut)]
-		let mut max_field: u64 = 0;
-		$(
-			if $type >= max_field { max_field = $type + 1; }
-		)*
 		$(
-			if $optional_type >= max_field { max_field = $optional_type + 1; }
+			encode_a_tlv!($stream, $type, $field, $fieldty);
 		)*
-		#[allow(unused_variables)]
-		for i in 0..max_field {
-			$(
-				if i == $type {
-					BigSize($type).write($stream)?;
-					BigSize($field.serialized_length() as u64).write($stream)?;
-					$field.write($stream)?;
-				}
-			)*
-			$(
-				if i == $optional_type {
-					if let Some(ref field) = $optional_field {
-						BigSize($optional_type).write($stream)?;
-						BigSize(field.serialized_length() as u64).write($stream)?;
-						field.write($stream)?;
-					}
-				}
-			)*
-		}
 	} }
 }
 
+macro_rules! get_a_varint_length_prefixed_tlv_length {
```

Comment:
Perhaps this could use a name that differs more from `get_varint_length_prefixed_tlv_length`. Maybe replace `get_a` with `calculate` since this macro does't evaluate to any value unlike `get_varint_length_prefixed_tlv_length`.

Reply:
Actually, can we collapse the two? I don't think there's any reason to have a separate `get_varint_length_prefixed_tlv_length` anymore, it can just be inlined in `encode_varint_length_prefixed_tlv`.

Reply:
I'm fine with inlining `get_varint_length_prefixed_tlv_length`, but note it'll also be inlined in a `serialized_length` implementation

Reply:
It's like three lines of code, I think that's fine.